File: python/examples/pn/04-mix/main_coop.py
"""PN 04-mix — cooperative multi-rate scheduler.

Same scenario as ``main.py`` but driven by ``CoopExecutive``.

``CoopExecutive`` differs from ``CyclicExecutive``:

- **Dynamic dispatch**: runs the runtime when its deadline passes rather
  than advancing a fixed slot table.
- **Overrun-tolerant**: deadline advances by one period regardless of
  actual execution time, preventing accumulated phase slippage.
- **Multi-runtime**: can drive several independent runtimes (useful when
  PN and FSM nodes cannot share a runtime).

All four nodes (light_a 10 ms, blink_b 10 ms, auto_c 20 ms, cli 10 ms)
are in a single runtime; ``CoopExecutive`` drives it at the base period
(10 ms).

Usage::

    cd python
    uv run examples/pn/04-mix/main_coop.py
"""
from __future__ import annotations

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[4]))
examples_root = str(Path(__file__).resolve().parents[2])
sys.path.insert(0, examples_root)
sys.path.insert(0, str(Path(examples_root) / "pn" / "01-light"))
sys.path.insert(0, str(Path(examples_root) / "pn" / "02-auto"))
sys.path.insert(0, str(Path(examples_root) / "pn" / "03-blink"))

from rxnet.coop import CoopExecutive
from rxnet.pn import Net, Runtime as PnRuntime
from rxnet.runtime import Context

import app_driver
from cli import Cli
from light_pn import P_ON as LIGHT_P_ON, create_light_pn
from auto_pn import P_ON as AUTO_P_ON, create_auto_pn, get_timeout_ms, set_timeout_ms
from blink_pn import P_X1, P_X2, create_blink_pn, get_base_hz, get_output_enabled, set_base_hz

LIGHT_A_GPIO = 2
LIGHT_B_GPIO = 4
LIGHT_C_GPIO = 5
BUTTON_A_GPIO = 0
BUTTON_B_GPIO = 15
DEFAULT_FREQ_B_HZ = 2
DEFAULT_TIMEOUT_C_MS = 9000

FAST_PERIOD_US = 10_000
SLOW_PERIOD_US = 20_000
CLI_PERIOD_US  = 10_000


class CliNode:
    def __init__(self, cli: Cli) -> None:
        self._cli = cli

    def latch_inputs(self, ctx: Context) -> None:
        pass

    def evaluate(self, ctx: Context) -> None:
        pass

    def commit(self, ctx: Context) -> None:
        pass

    def dump_outputs(self, ctx: Context) -> None:
        self._cli.tick()


def _light_state(net: Net, p_on: int) -> str:
    return "ON" if net.places[p_on] > 0 else "OFF"


def _blink_state(net: Net) -> str:
    if net.places[P_X2] > 0:
        return "X2"
    if net.places[P_X1] > 0:
        return "X1"
    return "OFF"


def _blink_effective_hz(net: Net) -> int:
    base = get_base_hz(net)
    if net.places[P_X2] > 0:
        return base * 2
    if net.places[P_X1] > 0:
        return base
    return 0


def cmd_button_a(line: str, user: object) -> None:
    if app_driver.trigger_button(BUTTON_A_GPIO):
        print("button A queued")
    else:
        print("button A trigger failed")


def cmd_button_b(line: str, user: object) -> None:
    if app_driver.trigger_button(BUTTON_B_GPIO):
        print("button B queued")
    else:
        print("button B trigger failed")


def cmd_status(line: str, nets: tuple[Net, Net, Net]) -> None:
    light_a, blink_b, auto_c = nets
    print(
        f"A(light): {_light_state(light_a, LIGHT_P_ON)}"
        f" | B(blink): {_blink_state(blink_b)} output={int(get_output_enabled(blink_b))}"
        f" | C(auto): {_light_state(auto_c, AUTO_P_ON)}"
    )
    print(
        f"B base_hz={get_base_hz(blink_b)} effective={_blink_effective_hz(blink_b)}"
        f" | C auto-off={get_timeout_ms(auto_c)} ms"
    )


def cmd_freq(line: str, nets: tuple[Net, Net, Net]) -> None:
    _, blink_b, _ = nets
    parts = line.split()
    if len(parts) != 2:
        print("usage: freq <hz>")
        return
    try:
        freq_hz = int(parts[1])
    except ValueError:
        print("usage: freq <hz>")
        return
    if freq_hz <= 0:
        print("frequency must be > 0 hz")
        return
    if set_base_hz(blink_b, freq_hz):
        print(f"blink B base frequency set to {freq_hz} hz")
    else:
        print("failed to update blink frequency (B)")


def cmd_timeout(line: str, nets: tuple[Net, Net, Net]) -> None:
    _, _, auto_c = nets
    parts = line.split()
    if len(parts) != 2:
        print("usage: timeout <ms>")
        return
    try:
        timeout_ms = int(parts[1])
    except ValueError:
        print("usage: timeout <ms>")
        return
    if timeout_ms <= 0:
        print("timeout must be > 0 ms")
        return
    if set_timeout_ms(auto_c, timeout_ms):
        print(f"auto C timeout set to {timeout_ms} ms")
    else:
        print("failed to update auto timeout (C)")


def cmd_quit(line: str, user: object) -> None:
    print("bye")
    sys.exit(0)


def main() -> None:
    light_a = create_light_pn(BUTTON_A_GPIO, LIGHT_A_GPIO)
    blink_b = create_blink_pn(BUTTON_A_GPIO, LIGHT_B_GPIO, DEFAULT_FREQ_B_HZ)
    auto_c  = create_auto_pn(BUTTON_B_GPIO, LIGHT_C_GPIO, DEFAULT_TIMEOUT_C_MS)

    cli = Cli()
    nets = (light_a, blink_b, auto_c)

    cli.register("a",        cmd_button_a)
    cli.register("press a",  cmd_button_a)
    cli.register("b",        cmd_button_b)
    cli.register("press b",  cmd_button_b)
    cli.register("status",   cmd_status,  nets)
    cli.register("freq",     cmd_freq,    nets)
    cli.register("timeout",  cmd_timeout, nets)
    cli.register("help",     lambda l, u: cli.print_help())
    cli.register("quit",     cmd_quit)
    cli.register("exit",     cmd_quit)
    cli.add_help_line("freq <hz>")
    cli.add_help_line("timeout <ms>")

    cli_node = CliNode(cli)

    rt = PnRuntime()
    rt.add_net(light_a,  FAST_PERIOD_US)
    rt.add_net(blink_b,  FAST_PERIOD_US)
    rt.add_net(auto_c,   SLOW_PERIOD_US)
    rt.add_node(cli_node, CLI_PERIOD_US)

    cli.print_help()
    cmd_status("status", nets)
    cli.print_prompt()

    ce = CoopExecutive()
    ce.add(rt)
    ce.run()  # never returns


if __name__ == "__main__":
    main()

File: c/examples/pn/04-mix/main_coop.c
// Copyright 2026 Jose M. Moya <jm.moya@upm.es>
// SPDX-License-Identifier: GPL-3.0-or-later

/*
 * PN 03-mix — cooperative (single-thread) multi-rate scheduler.
 *
 * Each net (and the CLI machine) carries its own period; rx_coop_exec
 * tracks the next activation instant for every periodic node:
 *
 *   pn_rt:  light_a 10 ms, blink_b 10 ms, auto_c 20 ms  [rx_pn_runtime]
 *             next activations at 10 ms / 20 ms periods
 *   cli_rt: cli     10 ms                                 [rx_fsm_runtime]
 *
 * rx_coop_exec runs whichever runtime is due, then sleeps until the
 * nearest upcoming deadline — no preemption, no OS threads.
 *
 * Contrast with main_cli.c (cyclic executive):
 *   - Cyclic exec: pre-computed static dispatch table.
 *   - Coop exec: dynamic deadline scheduling; tasks that run long do
 *     not corrupt the table — the scheduler just catches up on the
 *     next iteration.
 *
 * Contrast with main_threads.c:
 *   - Single thread: no concurrent access, no mutexes needed.
 *   - Suitable for cooperative RTOS (e.g. FreeRTOS taskYIELD pattern).
 *
 * Build:
 *   make -C c pn_03_mix_coop
 *   ./c/build/pn_03_mix_coop
 */

#include <stdio.h>
#include <stdlib.h>

#include "rxnet/coop.h"
#include "rxnet/fsm.h"
#include "rxnet/pn.h"

#include "app_driver.h"
#include "cli_fsm.h"
#include "light_pn.h"
#include "auto_pn.h"
#include "blink_pn.h"
#include "sched_report.h"

#define LIGHT_A_GPIO         2
#define LIGHT_B_GPIO         4
#define LIGHT_C_GPIO         5
#define BUTTON_A_GPIO        0
#define BUTTON_B_GPIO        15
#define FAST_PERIOD_US       10000L
#define SLOW_PERIOD_US       20000L
#define CLI_PERIOD_US        10000L
#define DEFAULT_FREQ_B_HZ    2u
#define DEFAULT_TIMEOUT_C_MS 9000u

typedef struct {
    rx_pn_net *light_a;
    rx_pn_net *blink_b;
    rx_pn_net *auto_c;
} app_data;

/* ------------------------------------------------------------------ */
/* CLI commands                                                         */
/* ------------------------------------------------------------------ */

static const char *light_state(const rx_pn_net *net, int p_on) {
    return (net != NULL && net->places[p_on] > 0) ? "ON" : "OFF";
}

static const char *blink_state(const rx_pn_net *net) {
    if (net == NULL) return "OFF";
    if (net->places[2] > 0) return "X2";
    if (net->places[1] > 0) return "X1";
    return "OFF";
}

static void
cmd_button_a(rx_fsm_context *ctx, cli_machine_data *cli,
             const char *command, void *ud)
{
    (void)ctx; (void)cli; (void)command; (void)ud;
    if (app_driver_trigger_button(BUTTON_A_GPIO) == ESP_OK) { printf("button A queued\n"); return; }
    printf("button A trigger failed\n");
}

static void
cmd_button_b(rx_fsm_context *ctx, cli_machine_data *cli,
             const char *command, void *ud)
{
    (void)ctx; (void)cli; (void)command; (void)ud;
    if (app_driver_trigger_button(BUTTON_B_GPIO) == ESP_OK) { printf("button B queued\n"); return; }
    printf("button B trigger failed\n");
}

static void
cmd_status(rx_fsm_context *ctx, cli_machine_data *cli,
           const char *command, void *ud)
{
    const app_data *app = (const app_data *)ud;
    (void)ctx; (void)cli; (void)command;
    printf(
        "A(light): %s | B(blink): %s output=%d | C(auto): %s\n",
        light_state(app->light_a, 1),
        blink_state(app->blink_b),
        blink_pn_get_output_enabled(app->blink_b),
        light_state(app->auto_c, 1)
    );
    printf(
        "B base_hz=%u | C auto-off=%u ms\n",
        blink_pn_get_base_hz(app->blink_b),
        auto_pn_get_timeout_ms(app->auto_c)
    );
}

static void
cmd_freq(rx_fsm_context *ctx, cli_machine_data *cli,
         const char *command, void *ud)
{
    app_data *app = (app_data *)ud;
    unsigned int freq_hz = 0u;
    (void)ctx; (void)cli;
    if (sscanf(command, "freq %u", &freq_hz) != 1) { printf("usage: freq <hz>\n"); return; }
    if (freq_hz == 0u) { printf("frequency must be > 0 hz\n"); return; }
    if (blink_pn_set_base_hz(app->blink_b, freq_hz) != 0) { printf("failed\n"); return; }
    printf("blink B base frequency set to %u hz\n", freq_hz);
}

static void
cmd_timeout(rx_fsm_context *ctx, cli_machine_data *cli,
            const char *command, void *ud)
{
    app_data *app = (app_data *)ud;
    unsigned int timeout_ms = 0u;
    (void)ctx; (void)cli;
    if (sscanf(command, "timeout %u", &timeout_ms) != 1) { printf("usage: timeout <ms>\n"); return; }
    if (timeout_ms == 0u) { printf("timeout must be > 0 ms\n"); return; }
    if (auto_pn_set_timeout_ms(app->auto_c, timeout_ms) != 0) { printf("failed\n"); return; }
    printf("auto C timeout set to %u ms\n", timeout_ms);
}

static void
cmd_help(rx_fsm_context *ctx, cli_machine_data *cli,
         const char *command, void *ud)
{
    size_t i;
    (void)ctx; (void)command; (void)ud;
    printf("Commands:\n");
    for (i = 0; i < cli->command_count; ++i)
        printf(" - %s\n", cli->commands[i].name);
    printf(" - freq <hz>\n - timeout <ms>\n");
}

static void
cmd_quit(rx_fsm_context *ctx, cli_machine_data *cli,
         const char *command, void *ud)
{
    (void)ctx; (void)cli; (void)command; (void)ud;
    printf("bye\n");
    exit(0);
}

/* ------------------------------------------------------------------ */
/* Main                                                                 */
/* ------------------------------------------------------------------ */

int
main(void)
{
    rx_pn_net light_a, blink_b, auto_c;
    rx_pn_runtime pn_rt;
    rx_fsm_runtime cli_rt;
    rx_fsm_machine cli_machine;
    app_data app = {&light_a, &blink_b, &auto_c};
    cli_machine_data cli_data;
    rx_coop_exec ce;
    rx_example_coop_sched_command sched_cmd = {"coop", &ce};

    if (rx_pn_runtime_init(&pn_rt,  3) != 0 ||
        rx_fsm_runtime_init(&cli_rt, 1) != 0) {
        fprintf(stderr, "runtime_init failed\n");
        return 1;
    }

    if (light_pn_init(&light_a, BUTTON_A_GPIO, LIGHT_A_GPIO) != 0 ||
        blink_pn_init(&blink_b, BUTTON_A_GPIO, LIGHT_B_GPIO, DEFAULT_FREQ_B_HZ) != 0 ||
        auto_pn_init(&auto_c,   BUTTON_B_GPIO, LIGHT_C_GPIO, DEFAULT_TIMEOUT_C_MS) != 0) {
        fprintf(stderr, "pn_init failed\n");
        return 1;
    }

    if (rx_pn_runtime_add_net(&pn_rt, &light_a, FAST_PERIOD_US, 0) != 0 ||
        rx_pn_runtime_add_net(&pn_rt, &blink_b, FAST_PERIOD_US, 0) != 0 ||
        rx_pn_runtime_add_net(&pn_rt, &auto_c,  SLOW_PERIOD_US, 0) != 0) {
        fprintf(stderr, "add_net failed\n");
        return 1;
    }

    cli_fsm_data_init(&cli_data, &app);
    if (cli_fsm_register_command(&cli_data, "a",        cmd_button_a, &app) != 0 ||
        cli_fsm_register_command(&cli_data, "press a",  cmd_button_a, &app) != 0 ||
        cli_fsm_register_command(&cli_data, "b",        cmd_button_b, &app) != 0 ||
        cli_fsm_register_command(&cli_data, "press b",  cmd_button_b, &app) != 0 ||
        cli_fsm_register_command(&cli_data, "status",   cmd_status,   &app) != 0 ||
        cli_fsm_register_command(&cli_data, "freq",     cmd_freq,     &app) != 0 ||
        cli_fsm_register_command(&cli_data, "timeout",  cmd_timeout,  &app) != 0 ||
        cli_fsm_register_command(&cli_data, "sched",    rx_example_cmd_coop_sched, &sched_cmd) != 0 ||
        cli_fsm_register_command(&cli_data, "help",     cmd_help,     NULL) != 0 ||
        cli_fsm_register_command(&cli_data, "quit",     cmd_quit,     NULL) != 0 ||
        cli_fsm_register_command(&cli_data, "exit",     cmd_quit,     NULL) != 0) {
        fprintf(stderr, "register_command failed\n");
        return 1;
    }
    cli_fsm_create(&cli_machine, "cli", &cli_data);

    if (rx_fsm_runtime_add_machine(&cli_rt, &cli_machine, CLI_PERIOD_US, 0) != 0) {
        fprintf(stderr, "add_machine cli failed\n");
        return 1;
    }

    cmd_help(&cli_rt.context, &cli_data, "help", NULL);
    cmd_status(&cli_rt.context, &cli_data, "status", &app);
    cli_fsm_print_prompt(&cli_data);

    rx_coop_exec_init(&ce);
    rx_coop_exec_add(&ce, &pn_rt.runtime);   /* light_a + blink_b + auto_c */
    rx_coop_exec_add(&ce, &cli_rt.runtime);  /* cli */
    rx_coop_exec_enable_sched_check(&ce, 1);
    rx_coop_exec_run(&ce); /* never returns */

    return 0;
}
